File: 10_gpu_speedup/problems/02_gpu_matmul/gpu_matmul.cpp
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <omp.h>

int main(int argc, char ** argv) {
  long n = (argc > 1 ? atol(argv[1]) : 1024);
  double * A = (double *)malloc(sizeof(double) * n * n);
  double * B = (double *)malloc(sizeof(double) * n * n);
  double * C = (double *)malloc(sizeof(double) * n * n);
  for (long i = 0; i < n * n; i++) { A[i] = 1.0; B[i] = 1.0; C[i] = 0.0; }

  double t0 = omp_get_wtime();
  /* 行列積 C = A * B (完成済み).
     OMP_TARGET_OFFLOAD=DISABLED ならホスト(CPU)で, MANDATORY ならGPUで実行される. */
#pragma omp target teams distribute parallel for map(to: A[0:n*n], B[0:n*n]) map(from: C[0:n*n])
  for (long i = 0; i < n; i++) {
    for (long j = 0; j < n; j++) {
      double s = 0.0;
      for (long k = 0; k < n; k++) s += A[i * n + k] * B[k * n + j];
      C[i * n + j] = s;
    }
  }
  double t1 = omp_get_wtime();
  double dt = t1 - t0;

  /* 検算: A,B が全て 1 なので C[i][j] = n になるはず */
  long err = 0;
  for (long i = 0; i < n * n; i++) if (C[i] != (double)n) err++;

  double flops = 2.0 * (double)n * (double)n * (double)n;
  printf("n = %ld, elapsed = %.3f sec, %.3f GFLOPS, %s\n",
         n, dt, flops / dt * 1e-9, (err == 0 ? "OK" : "NG"));
  free(A); free(B); free(C);
  return 0;
}

File: 10_gpu_speedup/problems/02_gpu_matmul/gpu_matmul.f90
program gpu_matmul
  use omp_lib
  character(len=32) :: arg
  integer(8) :: n, i, j, k, err
  real(8), allocatable :: A(:,:), B(:,:), C(:,:)
  real(8) :: s, t0, t1, dt, flops
  n = 1024
  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read (arg, *) n
  end if
  allocate(A(n,n), B(n,n), C(n,n))
  A = 1.0d0; B = 1.0d0; C = 0.0d0

  t0 = omp_get_wtime()
  ! 行列積 C = A * B (完成済み).
  ! OMP_TARGET_OFFLOAD=DISABLED ならホスト(CPU)で, MANDATORY ならGPUで実行される.
  !$omp target teams distribute parallel do map(to: A, B) map(from: C) private(j, k, s)
  do i = 1, n
     do j = 1, n
        s = 0.0d0
        do k = 1, n
           s = s + A(i,k) * B(k,j)
        end do
        C(i,j) = s
     end do
  end do
  !$omp end target teams distribute parallel do
  t1 = omp_get_wtime()
  dt = t1 - t0

  ! 検算: A,B が全て 1 なので C(i,j) = n になるはず
  err = 0
  do j = 1, n
     do i = 1, n
        if (C(i,j) /= dble(n)) err = err + 1
     end do
  end do

  flops = 2.0d0 * dble(n) * dble(n) * dble(n)
  if (err == 0) then
     print "(a,i0,a,f0.3,a,f0.3,a)", "n = ", n, ", elapsed = ", dt, " sec, ", flops / dt * 1.0d-9, " GFLOPS, OK"
  else
     print "(a,i0,a,f0.3,a,f0.3,a)", "n = ", n, ", elapsed = ", dt, " sec, ", flops / dt * 1.0d-9, " GFLOPS, NG"
  end if
end program gpu_matmul
